proxy/api/src/service.rs

```diff
@@ -4,13 +4,43 @@ use futures::prelude::*;
 use std::sync::Arc;
 use tokio::sync::{mpsc, Notify};
 
-#[derive(Clone)]
 /// Persistent configuration for running the API and coco peer services.
 pub struct Config {
     /// Secret key for the coco peer.
     ///
     /// If this is `None` coco is not started.
     pub key: Option<coco::keys::SecretKey>,
+    /// If set, we use a temporary directory for on-disk persistence.
+    pub temp_dir: Option<tempfile::TempDir>,
+}
+
+/// Error returned when constructing a new configuration
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// Failed to create temporary directory
+    #[error("Failed to create temporary directory")]
+    TempDir(
+        #[source]
+        #[from]
+        std::io::Error,
+    ),
+}
+
+impl Config {
+    /// Create a new default configuration.c:w
+    ///
+    /// If `test_mode` is `true` then `Config::temp_dir` is set for temporary on-disk persistence.
+    fn new(test_mode: bool) -> Result<Self, Error> {
```

Comment:
Might I suggest passing the `Option<TempDir>` as the argument here and forcing the layer above to handle determine test mode and creation of the directory?

With that change, I think resetting it below would consist of:
```rust
self.config = Config::new(self.config.temp_dir.map(|_| tempfile::tempdir()).transpose()?)
```

Reply:
You make a good point about moving the decision whether to construct the temp dir from `Config` to one level above. One problem that I see with your approach is that the construction of `self.config` is not paralleled between `Manager::new` and the reset message and the later is harder to understand. One thing that I was considering was moving he `Config::new` logic to `Manager::new_config`. Of course, this would need to be a static method because we need to have `config` before we can construct `Manager`.

Another thing to consider is to rename `Config` to something else (maybe `Environment`) which would make it less strange that it cannot be constructed without side-effects and failures.

Any thoughts @FintanH?

Reply:
> One problem that I see with your approach is that the construction of self.config is not paralleled between Manager::new and the reset message and the later is harder to understand

Can you explain a bit more what you mean here? 

> Another thing to consider is to rename Config to something else (maybe Environment) which would make it less strange that it cannot be constructed without side-effects and failures.

Ya this seems like a nicer move imo. It lends itself to the concept of controlling things, where the test mode flag controls how we do the set up :+1:

Reply:
> > One problem that I see with your approach is that the construction of self.config is not paralleled between Manager::new and the reset message and the later is harder to understand
> 
> Can you explain a bit more what you mean here?

The `Config` value we construct in `Manager::new()` should be the same value as the one constructed when handling `Message::Reset` in `Manager::config()`. This is what resetting the service config means. In both cases we want a pristine value determined by some external parameters (in our case only `test_mode`). To make this clear in the code and more robust against changes to how a pristine `Config` value is constructed it’s beneficial to construct it with a simple functional call that can be placed in both locations. The downside of your proposal is that constructing `Config` is very dissimilar in the two cases but yields the same result.

> > Another thing to consider is to rename Config to something else (maybe Environment) which would make it less strange that it cannot be constructed without side-effects and failures.
> 
> Ya this seems like a nicer move imo. It lends itself to the concept of controlling things, where the test mode flag controls how we do the set up +1

Do you have a suggestion for a term? I find `Environment` to be ok, but it’s not hitting it on the nail.

Reply:
How about `Env`? :smile:

Reply:
@FintanH: I’ve renamed `service::Config` to `Environment`. Could you have another look?

Reply:
@geigerzaehler would it make more sense to have the `test_mode` a field of the `Environment` rather than the `Manager`?

Reply:
That’s a good idea actually. It would allow us to drop the `args` argument from `run_rig`.